feat: sync all the things

antmat opened the pull request:
yoohooo

3Hren committed:
fix(logging): fix segmentation violation

Copy severity variable into the filter function to avoid out of the
scope variable usage.

3Hren committed:
version: bump - 0.12.6.3+nmu1

3Hren committed:
Merge branch 'v0.12.6' into develop

antmat committed:
refactor(errors): use cocaine::make_error_code instead of std

3Hren committed:
Merge pull request #72 from antmat/error_fix

refactor(errors): use cocaine::make_error_code instead of std

3Hren committed:
version: bump - 0.12.7.1+nmu1

antmat committed:
feat(config): several plugin paths can be specified now

3Hren committed:
Merge branch 'develop' into v0.12.7

3Hren committed:
Merge pull request #73 from antmat/multiple_paths

feat(config): several plugin paths can be specified now

3Hren committed:
version: bump - 0.12.7.2+nmu1

3Hren committed:
fix(): add missing include

3Hren committed:
version: bump - 0.12.7.3+nmu1

3Hren committed:
feat(logging): tracing logs as debug

3Hren committed:
changelog: update

3Hren committed:
Merge pull request #74 from 3Hren/issue/xx/tracing-severity

Tracing logs as debug

3Hren committed:
fix(runtime): fix invalid logging pattern

3Hren committed:
changelog: update

3Hren committed:
fix(logging): fix missing filter change

3Hren committed:
fix(version): bump release version

antmat committed:
refactor(unicorn): use std::function with futures instead of interfaces

3Hren committed:
Merge pull request #76 from antmat/unicorn_api_refactoring

refactor(unicorn): use std::function with futures instead of interfaces

antmat committed:
refactor(context): change context to pimpl

antmat committed:
fix(style): idents, perf on log filtering

antmat committed:
refactor(compile time): optimize includes, split hpack

antmat committed:
refactor(context): drop invoke and listen

drop invoke and listen in context_t in favour of signal_hub() method for
retroactive signals retreival

antmat committed:
refactor(context): context is interface now

antmat committed:
Merge pull request #77 from antmat/context_pimpl

refactor(context): change context to pimpl

3Hren committed:
fix(precise): fix build on precise

3Hren committed:
fix(): add missing include

antmat committed:
refactor(format): drop boost format in favor of BH formatter (cppformat)

* refactor(format): drop boost format in favor of BH formatter (cppformat)

* fix(format): use format function instead of writer usage

antmat committed:
Reduce compile time

* refactor(): granulated includes - reduce compile time
* refactor(): get rid of karma, use boost:join instead
* refactor(errors): move mutex inside registrar impl
* fix(build): add missing include after rebase
* fix(compile): add forgotten files
* fix(style): indent

antmat committed:
refactor(config): config is interface now

* refactor(config): config is interface now
* refactor(WIP): work on config_t interface
* fix(build): building now
* refactor(config): cleanup, build fix
* fix(config): add proper virtual destructors
* fix(style): rename visit to apply, rename config_impl_t
* fix(style): add comments, rename apply to each

antmat committed:
refactor(context): provide custom filter capability

* refactor(context): provide custom filter capability
* fix(perf): decrease time under lock

antmat committed:
fix(build): capture by reference to remove compiler error

antmat committed:
refactor(errors): hard code error categories instead of hashing

* refactor(errors): use well defined algo for hashing instead of std::hash
* refactor(registart): add registration with provided id
* refactor(error category): drop hashing support
* fix(style): drop unused variable

antmat committed:
Add generic category (#84)

* fix(logging): unregister any filter before terminate

As it can be bound in plugin, which is going to be unloaded

* fix(error): add generic category as well to category map.

antmat committed:
fix(build): addmissing includes, use config interface

antmat committed:
version(): bump

antmat committed:
fix(build): allow native-framework to build properly

antmat committed:
version(): bump

antmat committed:
fix(build): add missing header

antmat committed:
version(): bump

antmat committed:
feat(logging): more verbose error message

antmat committed:
version(): bump

antmat committed:
Do not throw on duplicate category insertion if they are really the same (#85)

* fix(error_category): do not throw on duplicate same category
* version(): bump

antmat committed:
Debianization fix (#86)

* fix(debian): parallel build. disabled init script by default

* fix(debian): restore cdbs, use parallel build

* fix(init): missing semicolon

* fix(debian): drop parallel build

antmat committed:
fix(locator): always link node (#87)

antmat committed:
version(): bump 0.12.8.5

antmat committed:
fix(session): prevent race in session::pull

antmat committed:
version(): bump 0.12.8.6

antmat committed:
fix(hpack): misprint - call method instead of variable use

antmat committed:
version(): bump 0.12.8.7

antmat committed:
fix(hpack): do not use shadowed variable

antmat committed:
version(): bump 0.12.8.8

antmat committed:
fix(hpack): add proper constructor for header_storage_t

antmat committed:
fix(hpack): proper ctor for header_storage_t

antmat committed:
version(): bump 0.12.8.9

3Hren committed:
fix(error): properly initialize static objects

3Hren committed:
version: bump - 0.12.8.10+nmu1

antmat committed:
fix(hpack): point header in table to proper place, tests. (#90)

antmat committed:
version(): bump 0.12.8.11

antmat committed:
Shared util lib (#91)

* refactor(util): make util shared library to prevent duplicated load

antmat committed:
version(bump): 0.12.8.12

antmat committed:
fix(category): compare error categories by name, not by pointer

antmat committed:
version(): bump 0.12.8.13

antmat committed:
fix(error_categories): really map by name, not by pointer (#92)

antmat committed:
version(): bump 0.12.8.14

3Hren committed:
feat: forward headers into slots

3Hren committed:
fix: message_queue now correctly saves headers

3Hren committed:
refactor: more forwards

3Hren committed:
refactor: rename meta to headers elsewhere

3Hren committed:
refactor: convenient naming

antmat committed:
fix(locator): do not close stream - it should be closed on read attempt

antmat committed:
Merge branch 'v0.12.8' into develop

3Hren committed:
refactor: drop no longer used friendship

3Hren committed:
feat: streamed<T> now throws if it was closed

3Hren committed:
refactor: rename context factory function

It better describes that it creates a new context without silent storing it somewhere.

3Hren committed:
refactor: move fields where they were before

3Hren committed:
refactor: Occam's razor

3Hren committed:
refactor: enable asserts, update unittests

3Hren committed:
refactor: drop deprecated invocation from slot

3Hren committed:
feat: add basic metrics to actors

New context functionality added to provide metrics hub, which can be used
for various metrics collection. For now actor connection stats added.

3Hren committed:
feat: add metrics for all service's events

3Hren committed:
refactor: adapt with Blackhole Release Candidate

3Hren committed:
refactor: depend on metrics for now

3Hren committed:
fix: fix build on GCC 4.8

3Hren committed:
version: bump - 0.12.9.0-alpha0+nmu1

3Hren committed:
fix: fixed SIGSEGV for client sessions

3Hren committed:
version: bump - 0.12.9.0-alpha1+nmu1

3Hren committed:
fix: disable assert

3Hren committed:
version: bump - 0.12.9.0-alpha2+nmu1

antmat committed:
fix(errors): more verbose error message

antmat committed:
fix(): proper category names comparision

3Hren committed:
fix: fix attributes duplication in logging service

3Hren committed:
version: bump - 0.12.9.0-alpha3+nmu1

antmat committed:
fix(): remove unused submodule

antmat committed:
fix(streamed): throw proper exception

antmat committed:
version(): bump 0.12.9.0-alpha4

antmat committed:
Statefull factory (#97)

* feat(repository): allow statefull factories for components

* refact(): statefull factories for repository

* fix(api): do not break API

antmat committed:
version(): bump v0.12.9.0-alpha5

antmat committed:
fix(hpack): correct find function and add tests (#99)

antmat committed:
refactor(context): drop useless and senseless terminate method (#98)

antmat committed:
version(): bump v0.12.9.1

3Hren committed:
feat: enable travis with Docker

3Hren committed:
chore: upgrade blackhole dependency (#101)

antmat committed:
refact(dlopen): use simple dlopen instead of ltdl - it provides better error messages (#102)

antmat committed:
Merge remote-tracking branch '3hren/v0.12.9' into HEAD

3Hren committed:
chore: github now has its own review system

3Hren committed:
fix: link with libdl (#105)

antmat committed:
Hpack refactoring (#104)

* refact(hpack): use owning headers

* fix(hpack): adpat tests to new api

* fix(): PR related fixes

* fix(): PR related fixes

* fix(): rename pack/unpack functions

* fix(): remove -O0

antmat committed:
refact(): storage now provides async operations. (#106)

3Hren committed:
version: bump - 0.12.10.1

3Hren committed:
fix: define proper release version

antmat committed:
fix: error by one in HPACK (#107)

3Hren committed:
fix: replace hardcode with hardcode

3Hren committed:
version: bump - 0.12.10.2

antmat committed:
fix: catch closed session exceptions (#108)

3Hren committed:
version: bump - 0.12.10.3

antmat committed:
fix: follow the HPACK RFC the way other pals do (#109)

antmat committed:
feat: custom component support (#111)

antmat committed:
refactor: proper naming for getting component group (#112)

3Hren committed:
version: bump - 0.12.10.4

antmat committed:
refactor: safe locate method for context (#113)

3Hren committed:
refactor: move here, move there

3Hren committed:
feat: introduce dispatch middlewares (#114)

- Add middleware concept.

Middleware represents a pre/post action performing around wrapped slot.
For example, an authorization component will be used as a middleware.

- Add iterative slot builder.

With this change you can now transparently configure service’s event
handlers without voodoo template magic. For example to add additional
meta_type argument to some handler as a first argument or to add just
introduced middlewares.

- Dispatch can now clear all slots registered using `halt` method.

3Hren committed:
feat: let slot builder register slots automatically (#115)

3Hren committed:
feat: more proper `result_of` implementation (#116)

3Hren committed:
fix: allow mutable methods to be deduced

3Hren committed:
refactor: adapt code with new `result_of`

3Hren committed:
refactor: newline at the EOF

3Hren committed:
refactor: pass event through middleware too (#117)

antmat committed:
refactor: allow to pass messages directly to upstream (#118)

antmat committed:
refactor: allow header packing reusage (#119)

Also fixed trace headers twice packing.

antmat committed:
fix: resolve name conflict (#120)

antmat committed:
refact(encoder): allow usage of encoded message outside of encoder (#122)

3Hren committed:
feat: dump gauges in metrics

3Hren committed:
version: bump - 0.12.10.5

3Hren committed:
Merge branch 'v0.12.10' into develop

3Hren committed:
feat: result_of improvements (#124)

* feat: result_of improvements

3Hren committed:
refactor: more forwards

antmat committed:
Vicodyn changes (#123)

* refact(vicodyn): vicodyn related minor refactoring
* refact(slot): introduce generic slot for custom protocols to avoid boilerplate

antmat committed:
refactor: make trace wrapper part of public API (#125)

3Hren committed:
fix: accept middleware by value (#126)

Its further usage requires that it will be copied anyway into the slot
builder.

By removing universal reference we fix improper type deduction for
lvalue references.

3Hren committed:
version: bump - 0.12.10.6

3Hren committed:
fix: forward all the things

3Hren committed:
version: bump - 0.12.10.7

3Hren committed:
Merge branch 'issue/forward' into develop

antmat committed:
refact(locator): move some logic inside gateway (#128)

* refact(locator): move some logic inside gateway

* fix(): code clarification

* fix(codestyle): clarify code a little

* refact(locator): make gateway optional again

* refact(cluster): add mode to prevent useless subscriptions

* refact(gateway): add resolve_policy

* fix(locator): move if to proper place

* refact(comments): add some notices about resolve_policy

antmat committed:
fix(build): specialize rebind for stupid GCC (#129)

3Hren committed:
version: bump - 0.12.11.0-alpha1

3Hren committed:
debian: proper dev dependency from metrics

3Hren committed:
ci: fix travis

antmat committed:
fix(ABI): restore ABI

antmat committed:
version: bump 0.12.11.0-alpha2

3Hren committed:
fix: exception safety for running actors

3Hren committed:
refactor: bump version in cmake file

3Hren committed:
version: bump - 0.12.11.0-alpha3

3Hren committed:
refactor: move common auth code from plugins

3Hren committed:
refactor: migrate even more code

3Hren committed:
refactor: migration is done

3Hren committed:
feat: bayan

3Hren committed:
version: bump - 0.12.11.0-alpha4

3Hren committed:
changelog: more verbose

3Hren committed:
fix: resolve remapped name instead of initial

This commit fixes a bug, where the locator resolved initial service
name instead of remapped one, which led to incorrect resolving for
remote routing groups.

3Hren committed:
chore: code style

3Hren committed:
version: bump - 0.12.11.0-alpha5

3Hren committed:
debfuck (#133)

* refactor: drop unnecessary public include

* refactor: indents

* debian: proper dev package dependencies

3Hren committed:
version: bump - 0.12.11.0-alpha6

antmat committed:
feat: executors (#134)

antmat committed:
feat: executor supports callbacks with move-only types

antmat committed:
version: bump v0.12.11.0-alpha7

antmat committed:
fix: build (#136)

3Hren committed:
version: bump - 0.12.11.0-beta1

antmat committed:
refact: better dynamic - print missing key (#138)

3Hren committed:
feat: access-controlled collections (#137)

This PR introduces a concept of secure storage collections.

By default all collections are considered common, and anyone is allowed to perform read or write operations. But after the first `write` operation a collection is captured by UIDs that performed such operation and becomes owned with `RW` access for UIDs that captured it and with no access to anyone else. This behaviour is handled by recording another secure collection, where all ACL's lay.

For example if an user `42` performed write operation over `secrets` collection and `who-killed-kennedy` key, the following record will be created at `<SECURE>` collection with key `secrets`:

```json
{
    "42": 3
}
```

That means that user `42` has now RW access (0x01 | 0x02) to collection `secrets`. That also means that it can change `<SECURE>` collection itself for keys where there is write access allowed, which It is useful for permission changes.

Moreover this PR adds a new category of logs - audit logs - with strict recording of **who**, **when**, **what** and **where** have been done with collections.

For example:
```
I, 2017-02-15 09:18:55.924: completed 'write' operation [size=22, event=write, collection=secrets, key=key, uids=[42], service=storage, source=audit]

W, 2017-02-15 09:18:55.922: failed to complete 'read' operation [code=2, error=[2] unauthorized, event=read, collection=collection, key=key, uids=[], service=storage, source=audit]
W, 2017-02-15 09:18:55.925: failed to complete 'write' operation [code=13, error=[13] Permission denied, event=write, collection=secrets, key=key, uids=[100500], service=storage, source=audit]
```

Currently all audit logs are written directly into the `core` logger, but this will be changed in further releases.

New component is introduced in this PR: controllers - things that verify access to a service's methods.

To configure `collection` controller, add these lines into the config:

```js
 "controllers": {
    "collection": {
        "type": "collection", // Also possible "null" value, that disables collection control.
        "args": {
            "backend": "core"
        }
    }
}
```

Features done:
- [x] Verbose logging.
- [x] Improved middleware support.
- [x] Audit logs.
- [x] Documentation.
- [x] Config example.

3Hren committed:
refactor: dead service elimination

3Hren committed:
version: bump - 0.12.11.1

3Hren committed:
fix: do not depend on zookeeper

3Hren committed:
refactor: minor actor changes

3Hren committed:
refactor: leave raw pointers to C-nerds

3Hren committed:
fix: catch exceptions while attaching node

3Hren committed:
version: bump - 0.12.11.2

3Hren committed:
fix: check whether all values are convertible

Not the first one.

3Hren committed:
refactor: preparation for Unicorn authorization

antmat committed:
refactor: migrate trace to channel (#142)

3Hren committed:
fix: properly convert dynamics into enums (#145)

3Hren committed:
version: bump - 0.12.11.3

3Hren committed:
fix: unhide virtual function

antmat committed:
fix: understandable exceptions in component instantiation (#146)

3Hren committed:
version: bump - 0.12.12.1

antmat committed:
feat: allow comments and commas in config (#147)

- Upgrade rapidjson, simplify parsing, and allow comments.
- Add `dynamic_t` construction from reference wrapper.
- Remove redundant code and warning in release build
- Hide rapidjson symbols in version script.

3Hren committed:
version: bump - 0.12.12.2

3Hren committed:
fix: specify uid type precisely

3Hren committed:
feat: anonymous access for storage authorization

3Hren committed:
ci: fix dockerfile

3Hren committed:
version: bump - 0.12.12.3

antmat committed:
refactor: better dynamic constructors

3Hren committed:
chore: code style

3Hren committed:
refactor: minor code swapping

3Hren committed:
chore: add a todo

3Hren committed:
feat: verbose logging using tracebit (#150)

A new HPACK header is now analyzed: a `trace_bit` of a boolean type.
Putting such header triggers verbose mode, when all logs are written with no severity check. This mimics the previous logic, when the same behaviour was achieved just by presence of a `trace_id` header.

antmat committed:
refactor: better streamed and deferred with error code instead of exceptions (#151)

antmat committed:
fix: build - gcc can not capture variadic pack

3Hren committed:
chore: make default config actual

3Hren committed:
debian: fix control

3Hren committed:
chore: typo fix

3Hren committed:
debian: get rid of `v12` infixes

3Hren committed:
refactor: dead code elimination

3Hren committed:
fix: add missing include

3Hren committed:
chore: comments style

3Hren committed:
refactor: dead code elimination

3Hren committed:
chore: code style

3Hren committed:
refactor: typo fix

3Hren committed:
refactor: no longer required to add json formatter

3Hren committed:
refactor: runtime cleanup (#153)

- Drop SFINAE magic in signal registering, because fuck you, that’s why.
- Added signal engine façade, which encapsulates signal handling
runtime and allows automatically join/cancel thread/cancellations.
- Fixed ignored logging backend parameter (always used «core» before).
- Replaced hardcoded return codes with defined ones.
- Minor renaming and includes elimination.

kobolog commented:
Чуваки, а можно как-нибудь сделать чтобы не было конфликтов, я б тогда помержил?

3Hren commented:
Попробую.

3Hren committed:
chore: typo fix

3Hren committed:
fix: make storage with authorization async again (#154)

Storage authorization check internally uses storage backend for
fetching/saving ACLs and it was implemented using synchronous storage
methods, which may freeze current execution’s thread. This commit
returns back asynchronous behaviour.

- Drop synchronous version of authorization method.
- Implement asynchronous version of authorization method.
- Add tests.

3Hren committed:
feat: combined storage authorization

3Hren committed:
refactor: merge branch 'cocaine/master' -> develop

3Hren commented:
Готово!

3Hren committed:
refactor: code deduplication

3Hren committed:
feat: extend format functionality

- Now `cocaine::format` can use newly introduced `display_traits`
specializations of  for fine-grained formatting.
- Fixed bug in vector displaying.

3Hren committed:
feat: add lvalue for this setters for identity builder

kobolog commented:
```
[  3%] Building CXX object cache/CMakeFiles/cache.dir/src/module.cpp.o
/build/cocaine-plugins/cache/src/module.cpp:23:21: error: expected type-specifier
     validation() -> api::preconditions_t {
                     ^
/build/cocaine-plugins/cache/src/module.cpp:23:21: error: expected initializer
/build/cocaine-plugins/cache/src/module.cpp:31:1: error: expected '}' at end of input
 }
 ^
```

Не похоже что готово, Евгений!

3Hren commented:
Значит не готово!

antmat commented:
это просто Тревис наложил санкции!

3Hren committed:
chore: code style

antmat committed:
refactor: better display traits (#158)

3Hren committed:
feat: combined event authorization (#157)

- Use both client & user id’s as authorization subjects.
- Allow to extend even more.

3Hren committed:
feat: protocol extension (#152)

This commit applies protocol extension frames similar to HTTP/2.

Implemented as a protocol extension frames, that allows sessions to
adjust their settings externally, revoke channels etc.

- Added new protocol - control.
- For all prime protocols new events are pushed back with event id
descending from 2**16-1. Such agreement guarantees that no one will be
broken during applying new scheme.
- The previous requirement forces us to use MPL map instead of type
lists for managing protocol event numbers.
- Transition states, such as `primitive<T>` or `streaming<T>` are now
required to be marked with `transition_type` to be properly handled by
cocaine’s machinery.
- New events are intentionally allowed for an end user to be called.
- Implemented channel revocation.

antmat committed:
refactor: allow upstream demolish session in emergency case (#156)

3Hren committed:
fix: restore used macro

kobolog commented:
Ну давайте уже сделайте чтобы собиралось, пора мержить!

antmat committed:
fix: specialize display_traits for attributes (#159)

This fixes build on GCC.

3Hren commented:
Да чет непонятная хуйня, у меня на компьютере все собирается!

3Hren committed:
changelog: update

3Hren committed:
debian: fix build-depends-on-1-revision

3Hren committed:
debian: upgrade standard version

3Hren committed:
refactor: code style

3Hren committed:
refactor: drop constness from dispatch

- No longer keep dispatch as const. It should mutate its state,
otherwise it’s useless!
- Drop `const` from prototypes.
- Drop `const` from `dispatch::process` and `dispatch::discard`.

3Hren committed:
changelog: update

3Hren committed:
refactor: eye-candy (#164)

3Hren committed:
feat: tag ACL collection (#163)

This allows to perform migrations in the future in the case of changed
format.

3Hren committed:
feat: allow context to insert with initializers (#162)

This is useful to atomically check-and-insert new services, especially
for Node service.

3Hren committed:
changelog: updae

3Hren committed:
fix: block signals before context creation

3Hren committed:
fix: pass proper arguments (#166)

3Hren committed:
refactor: authentication now throws (#167)

Instead of ADT.

3Hren committed:
fix: catch more exceptions (#165)

3Hren committed:
version: bump - 0.12.13.0-alpha1

3Hren commented:
Вот теперь все должно собираться!

noxiouz commented:
Yay!

3Hren committed:
refactor: unify ACL prefixes

3Hren committed:
version: bump - 0.12.13.0-alpha2

3Hren committed:
feat: build tests

- Add google test as a foreign submodule instead of on-demand
downloading.
- Enable tests.

3Hren committed:
refactor: minor uuid refactoring

antmat committed:
debian: strict dependency on io-util (#169)

antmat committed:
feat: map display traits (#170)

kobolog commented:
🥇

antmat commented:
much commits, such PR, wow!
